pkg/connector/functions.go

```diff
@@ -0,0 +1,164 @@
+package connector
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	"github.com/conductorone/baton-sdk/pkg/annotations"
+	"github.com/conductorone/baton-sdk/pkg/pagination"
+	rs "github.com/conductorone/baton-sdk/pkg/types/resource"
+	"github.com/conductorone/baton-segment/pkg/segment"
+)
+
+type functionResourceBuilder struct {
+	resourceType *v2.ResourceType
+	client       *segment.Client
+}
+
+func (f *functionResourceBuilder) ResourceType(_ context.Context) *v2.ResourceType {
+	return f.resourceType
+}
+
+// Create a new connector resource for an Segment Function.
+func functionResource(function *segment.Function, parentResourceID *v2.ResourceId) (*v2.Resource, error) {
+	resource, err := rs.NewResource(
+		function.DisplayName,
+		functionResourceType,
+		function.ID,
+		rs.WithParentResourceID(parentResourceID),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return resource, nil
+}
+
+func (f *functionResourceBuilder) List(ctx context.Context, parentResourceID *v2.ResourceId, pToken *pagination.Token) ([]*v2.Resource, string, annotations.Annotations, error) {
+	if parentResourceID == nil {
+		return nil, "", nil, nil
+	}
+
+	// There are 3 types of functions, we need to fetch all of them
+	functionTypes := []string{"DESTINATION", "INSERT_DESTINATION", "SOURCE"}
+	var cursor string
+	var allFunctions []segment.Function
+
+	for _, t := range functionTypes {
+		cursor = ""
+		for {
+			// Fetch data for the current type
+			functions, nextCursor, err := f.client.ListFunctions(ctx, cursor, t)
+			if err != nil {
+				return nil, "", nil, fmt.Errorf("error fetching data for type %s: %w", t, err)
+			}
+			allFunctions = append(allFunctions, functions...)
+
+			if nextCursor != "" {
+				cursor = nextCursor
+			} else {
+				// No more items for the current type, move to the next one
+				break
+			}
+		}
+	}
+
+	var rv []*v2.Resource
+	for _, fn := range allFunctions {
+		fnCopy := fn
+		fr, err := functionResource(&fnCopy, parentResourceID)
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		rv = append(rv, fr)
+	}
+
+	return rv, "", nil, nil
+}
+
+func (f *functionResourceBuilder) Entitlements(ctx context.Context, resource *v2.Resource, pToken *pagination.Token) ([]*v2.Entitlement, string, annotations.Annotations, error) {
+	bag, page, err := parsePageToken(pToken.Token, &v2.ResourceId{ResourceType: roleResourceType.Id})
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	roles, nextCursor, err := f.client.ListRoles(ctx, page)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	pageToken, err := bag.NextToken(nextCursor)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	var rv []*v2.Entitlement
+	for _, role := range roles {
+		if strings.Contains(role.Name, "Function") {
```

Comment:
In a previous loop (functions) we pin the iterator, but we don't here. Why is it an issue in the previous for loop but not now? Aren't these both pointers?